### PR TITLE
CASSANDRA-17819: Fix resetting the schema

### PR DESCRIPTION
**Pull request by jacek-lewandowski**

### CODE REVIEW

File: src/java/org/apache/cassandra/config/CassandraRelevantProperties.java [hunk 1]
```diff
@@ -235,6 +235,7 @@
     MEMTABLE_OVERHEAD_SIZE("cassandra.memtable.row_overhead_size", "-1"),
     MEMTABLE_OVERHEAD_COMPUTE_STEPS("cassandra.memtable_row_overhead_computation_step", "100000"),
     MIGRATION_DELAY("cassandra.migration_delay_ms", "60000"),
+    SCHEMA_PULL_INTERVAL("cassandra.schema_pull_interval_ms", "60000"),
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/config/CassandraRelevantProperties.java]: How about some Javadoc?

File: src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java [hunk 2]
```diff
@@ -232,12 +240,15 @@ public synchronized SchemaTransformationResult apply(SchemaTransformation transf
...
+        if (!update.diff.isEmpty())
         {
-            migrationCoordinator.announce(update.after.getVersion());
+            this.schema = update.after;
+            logger.debug("Schema updated: {}", update);
+            updateCallback.accept(update, true);
+            if (!local)
+            {
+                migrationCoordinator.announce(update.after.getVersion());
+            }
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java:264]: Shall we add some logging in case no update has happened?

File: src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java [hunk 3]
```diff
@@ -253,20 +264,28 @@ private synchronized SchemaTransformationResult reload()
...
-    public SchemaTransformationResult reset(boolean local)
+    public void reset(boolean local)
     {
         if (local)
-            return reload();
-
-        Collection<Mutation> mutations = migrationCoordinator.pullSchemaFromAnyNode().awaitThrowUncheckedOnInterrupt().getNow();
-        return applyMutations(mutations);
+        {
+            reload();
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java:287]: Do we still need to return in `reload()`? Seems like this is the only usage

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 4]
```diff
@@ -323,7 +372,7 @@ private boolean shouldPullFromEndpoint(InetAddressAndPort endpoint)
 
         if (!messagingService.versions.knows(endpoint))
         {
-            logger.debug("Not pulling schema from {} because their messaging version is unknown", endpoint);
+            logger.trace("Not pulling schema from {} because their messaging version is unknown", endpoint);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Why this was switched to trace?

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 5]
```diff
@@ -403,16 +461,61 @@ private synchronized void removeEndpointFromVersion(InetAddressAndPort endpoint,
...
+            }
+        });
+    }
+
     synchronized void removeAndIgnoreEndpoint(InetAddressAndPort endpoint)
     {
+        logger.debug("Removing and ignoring endpoint {}", endpoint);
         Preconditions.checkArgument(endpoint != null);
+        // TODO The endpoint address is now ignored but when a node with the same address is added again later,
+        //  there will be no way to include it in schema synchronization other than restarting each other node
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: I think I mentioned in the previous version but for completeness - I think this TO DO deserves to be documented actually

File: src/java/org/apache/cassandra/schema/Schema.java [hunk 6]
```diff
@@ -611,18 +613,42 @@ public SchemaTransformationResult transform(SchemaTransformation transformation,
      * Clear all locally stored schema information and fetch schema from another node.
      * Called by user (via JMX) who wants to get rid of schema disagreement.
      */
-    public synchronized void resetLocalSchema()
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/Schema.java:614]: Why this change?

File: src/java/org/apache/cassandra/schema/Schema.java [hunk 7]
```diff
@@ -611,18 +613,42 @@ public SchemaTransformationResult transform(SchemaTransformation transformation,
...
+    public void resetLocalSchema()
     {
         logger.debug("Clearing local schema...");
-        updateHandler.clear();
 
-        logger.debug("Clearing local schema keyspace instances...");
-        distributedKeyspaces.forEach(this::unload);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/Schema.java:620]: Where did the unload go?

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 8]
```diff
@@ -95,21 +98,37 @@ private void selectSilent(Cluster cluster, String name)
...
             assertTrue(cluster.get(1).callOnInstance(() -> Schema.instance.getTableMetadata(KEYSPACE, "tbl") != null));
             assertTrue(cluster.get(2).callOnInstance(() -> Schema.instance.getTableMetadata(KEYSPACE, "tbl") != null));
 
+            // now we have a table tbl in the schema of both nodes
+
             cluster.get(2).shutdown().get();
 
-            // when schema is removed and there is no other node to fetch it from, node 1 should be left with clean schema
-            cluster.get(1).runOnInstance(() -> Schema.instance.resetLocalSchema());
+            // when there is no node to fetch the schema from, reset local schema should immediately fail
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java:148]: How does this work in previous versions?

File: src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java [hunk 9]
```diff
@@ -253,20 +264,28 @@ private synchronized SchemaTransformationResult reload()
...
+            {
+                logger.error("Timeout exceeded when waiting for schema from other nodes");
+            }
+        }
     }
 
     @Override
     public synchronized void clear()
     {
+        schema = DistributedSchema.EMPTY;
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java]: Why did you decide to move it up?

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 10]
```diff
@@ -369,22 +418,31 @@ private synchronized boolean shouldApplySchemaFor(VersionInfo info)
...
 
         VersionInfo info = versionInfo.computeIfAbsent(version, VersionInfo::new);
         if (Objects.equals(schemaVersion.get(), version))
+        {
             info.markReceived();
+            logger.trace("Schema {} from {} has been marked as recevied because it is equal the local schema", version, endpoint);
+        }
         info.endpoints.add(endpoint);
-        info.requestQueue.addFirst(endpoint);
+        info.requestQueue.addFirst(endpoint); // TODO not sure if it is correct - given we've just marked this schema version as received, why do we add a request to receive it?
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Have to think a bit about this TO DO

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 11]
```diff
@@ -149,12 +153,27 @@ private static Set<InetAddressAndPort> getIgnoredEndpoints()
...
+        /**
+         * The set of endpoints containing this schema version
+         */
         final Set<InetAddressAndPort> endpoints           = Sets.newConcurrentHashSet();
+        /**
+         * The set of endpoints from which we are already fetching the schema
+         */
         final Set<InetAddressAndPort> outstandingRequests = Sets.newConcurrentHashSet();
+        /**
+         * The queue of endpoint from which we are going to fetch the schema
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: nit: endpoints

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 12]
```diff
@@ -180,6 +199,18 @@ boolean wasReceived()
         {
             return receivedSchema;
         }
+
+        @Override
+        public String toString()
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java:205]: Do we actually use this one anywhere?

**Review by ekaterinadimitrova2 — COMMENTED** [review]
Left some comments, primarily questions actually. Need to look at it more tomorrow. Seems like there are few things going on and I do not want to change any behavior

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java (hunk 8 above)
Comment (jacek-lewandowski) [reply test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java:148]: AFAIU it worked in the way that if we don't get a schema from some other node, we will be left in an inconsistent state, where we have `Keyspace` instances for keyspaces which are missing in the local schema. We would have to restart the node in such case. Also if some keyspaces were modified on the other node in the meantime, say a table was dropped - that table will not be dropped on the node were we called reset because from that node's view point, nothing has been dropped, and the stale data files will be left.

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java (hunk 12 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java:205]: in log messages

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java:205]: Yes, but I didn't see it used, probably I missed it :-)

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 13]
```diff
@@ -220,18 +251,23 @@ boolean wasReceived()
 
     void start()
     {
+        logger.info("Starting migration coordinator and scheduling pulling schema versions every 1 minute");
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: I think we need to make it print the SCHEMA_PULL_INTERVAL as it is mutable, not always 1 minute necessarily anymore

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 14]
```diff
@@ -426,49 +529,19 @@ private Future<Void> scheduleSchemaPull(InetAddressAndPort endpoint, VersionInfo
         FutureTask<Void> task = new FutureTask<>(() -> pullSchema(endpoint, new Callback(endpoint, info)));
 
         if (shouldPullImmediately(endpoint, info.version))
+        {
+            logger.debug("Pulling {} immediatelly from {}", info, endpoint);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: immediately

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 15]
```diff
@@ -403,16 +461,61 @@ private synchronized void removeEndpointFromVersion(InetAddressAndPort endpoint,
...
+
+        this.endpointVersions.clear();
+
+        Iterator<Map.Entry<UUID, VersionInfo>> it = versionInfo.entrySet().iterator();
+        while (it.hasNext())
+        {
+            Map.Entry<UUID, VersionInfo> entry = it.next();
+            it.remove();
+            entry.getValue().waitQueue.signal();
+        }
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java:488]: This feels like it deserves its own method?

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 16]
```diff
@@ -403,16 +461,61 @@ private synchronized void removeEndpointFromVersion(InetAddressAndPort endpoint,
...
+                if (state != null)
+                {
+                    UUID v = state.getSchemaVersion();
+                    if (v != null)
+                    {
+                        reportEndpointVersion(endpoint, v);
+                    }
+                }
+            }
+        });
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: This one too
Not that I am super opinionated but it just feels like it can improve readability that way

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 17]
```diff
@@ -95,25 +107,54 @@ private void selectSilent(Cluster cluster, String name)
...
             assertTrue(cluster.get(2).callOnInstance(() -> Schema.instance.getTableMetadata(KEYSPACE, "tbl") != null));
 
+            // now we have a table tbl in the schema of both nodes
+
             cluster.get(2).shutdown().get();
 
-            // when schema is removed and there is no other node to fetch it from, node 1 should be left with clean schema
-            cluster.get(1).runOnInstance(() -> Schema.instance.resetLocalSchema());
+            Awaitility.await().atMost(Duration.ofSeconds(30)) // wait until node 1 notices that node 2 is dead
+                      .until(() -> cluster.get(1).callOnInstance(() -> Gossiper.instance.getLiveMembers().stream().allMatch(e -> e.equals(FBUtilities.getBroadcastAddressAndPort()))));
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: I think this can be split in a few lines as it becomes too long

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 18]
```diff
@@ -95,25 +107,54 @@ private void selectSilent(Cluster cluster, String name)
...
+            cluster.get(1).runOnInstance(() -> {
+                Schema.instance.transform(current -> Schema.instance.distributedKeyspaces().without(KEYSPACE), false);
+             });
+
             assertTrue(cluster.get(1).callOnInstance(() -> Schema.instance.getTableMetadata(KEYSPACE, "tbl") == null));
 
+            // clear will wait until it receives schema from some other node
+            // also if we start the node2 first, schema of node2 will be synced to schema of node1 because node1 has the newest change (dropping the table)
+            // if we run clean on node1 first, it will advertise empty schema to node2 and apply the schema from it without merging
+            CompletableFuture<Boolean> clear1 = CompletableFuture.supplyAsync(() -> cluster.get(1).callOnInstance(() -> Schema.instance.updateHandler.clear().awaitUninterruptibly(1, TimeUnit.MINUTES)));
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: This line can also be split in a few lines so we don't have to scroll right

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 19]
```diff
@@ -95,25 +107,54 @@ private void selectSilent(Cluster cluster, String name)
...
+            // this proves node1 reset schema works - the most recent change should be discarded because it receives
+            // the schema from node2 and applies it on a clean schema
             Awaitility.waitAtMost(Duration.ofMinutes(1))
                       .pollDelay(Duration.ofSeconds(1))
                       .until(() -> cluster.get(1).callOnInstance(() -> Schema.instance.getTableMetadata(KEYSPACE, "tbl") != null));
 
+            // now let's break schema locally and let it be reset
+            cluster.get(1).runOnInstance(() -> Schema.instance.getLocalKeyspaces()
+                                                              .get(SchemaConstants.SCHEMA_KEYSPACE_NAME)
+                                                              .get().tables.forEach(t -> ColumnFamilyStore.getIfExists(t.keyspace, t.name).truncateBlockingWithoutSnapshot()));
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java:177]: This line also seems too long

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 20]
```diff
@@ -19,15 +19,27 @@
...
+import org.apache.cassandra.config.CassandraRelevantProperties;
+import org.apache.cassandra.db.ColumnFamilyStore;
 import org.apache.cassandra.distributed.Cluster;
 import org.apache.cassandra.distributed.api.ConsistencyLevel;
 import org.apache.cassandra.distributed.api.Feature;
+import org.apache.cassandra.gms.Gossiper;
 import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.SchemaConstants;
+import org.apache.cassandra.schema.SchemaKeyspace;
+import org.apache.cassandra.service.StorageService;
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: Unused imports

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 21]
```diff
@@ -19,15 +19,27 @@
...
+import org.apache.cassandra.utils.FBUtilities;
+import org.assertj.core.api.Assertions;
 import org.awaitility.Awaitility;
 
+import static org.apache.cassandra.schema.SchemaKeyspaceTables.ALL;
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: Unused import

**Review by ekaterinadimitrova2 — COMMENTED** [review]
Left some small comments, I want to do one more pass and check further the MigrationCoordinatorTest as I just skimmed through it. I want to be sure I didn't miss anything. I also want to look further in 4.0. Not an area of the code that I have a lot of experience

File: src/java/org/apache/cassandra/config/CassandraRelevantProperties.java (hunk 1 above)
Comment (adelapena) [inline src/java/org/apache/cassandra/config/CassandraRelevantProperties.java]: I'd add a brief JavaDoc line saying what the property does.

Comment (adelapena) [inline src/java/org/apache/cassandra/config/CassandraRelevantProperties.java]: Probably we should include the unit in the name, like in `SCHEMA_PULL_INTERVAL_MS`.

File: src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java [hunk 22]
```diff
@@ -85,8 +90,21 @@ public DefaultSchemaUpdateHandler(MigrationCoordinator migrationCoordinator,
...
+            {
+                if (requestedReset == null)
+                    applyMutations(msg.payload);
+            }
+        });
+        SchemaPullVerbHandler.instance.register(msg -> {
+            try
+            {
+                messagingService.send(msg.responseWith(getSchemaMutations()), msg.from());
+            } catch (RuntimeException ex) {
```
Comment (adelapena) [inline src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java]:
```suggestion
            }
            catch (RuntimeException ex)
            {
```

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 23]
```diff
@@ -23,24 +23,24 @@
...
+import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
-import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BiConsumer;
 import java.util.function.LongSupplier;
 import java.util.function.Supplier;
+import java.util.stream.Collectors;
```
Comment (adelapena) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Nit: unused import

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 24]
```diff
@@ -85,6 +85,10 @@
  * schema. It performs periodic checks and if there is a schema version mismatch between the current node and the other
  * node, it pulls the schema and applies the changes locally through the callback.
  *
+ * In particular the Migration Coordinator keeps track of all schema versions reported from each node in the cluster.
+ * As long as the certain version is advertised by some node, it is being tracked. As long as a version is tracked,
```
Comment (adelapena) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]:
```suggestion
 * As long as a certain version is advertised by some node, it is being tracked. As long as a version is tracked,
```

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java [hunk 25]
```diff
@@ -180,6 +199,18 @@ boolean wasReceived()
...
         }
+
+        @Override
+        public String toString()
+        {
+            return "VersionInfo{" +
+                   "version=" + version +
+                   ", outstandingRequests=" + outstandingRequests +
+                   ", requestQueue=" + requestQueue +
+                   ", waitQueue=" + waitQueue.getWaiting() +
```
Comment (adelapena) [inline src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Maybe it should be `waitQueueSize`?
```suggestion
                   ", waitQueueSize=" + waitQueue.getWaiting() +
```

File: src/java/org/apache/cassandra/schema/Schema.java [hunk 26]
```diff
@@ -611,18 +615,27 @@ public SchemaTransformationResult transform(SchemaTransformation transformation,
...
-        SchemaDiagnostics.schemaCleared(this);
+        if (Gossiper.instance.getLiveMembers().stream().allMatch(ep -> FBUtilities.getBroadcastAddressAndPort().equals(ep)))
+            throw new InvalidRequestException("Cannot reset local schema when there are no other live nodes");
 
-        updateHandler.reset(false);
-        logger.info("Local schema reset is complete.");
+        Awaitable clearCompletion = updateHandler.clear();
+        try
+        {
+            if (!clearCompletion.await(StorageService.SCHEMA_DELAY_MILLIS, TimeUnit.MILLISECONDS)) {
```
Comment (adelapena) [inline src/java/org/apache/cassandra/schema/Schema.java]:
```suggestion
            if (!clearCompletion.await(StorageService.SCHEMA_DELAY_MILLIS, TimeUnit.MILLISECONDS))
            {
```

File: src/java/org/apache/cassandra/schema/SchemaUpdateHandler.java [hunk 27]
```diff
@@ -63,14 +64,15 @@
...
      */
-    SchemaTransformationResult reset(boolean local);
+    void reset(boolean local);
 
     /**
      * Clears the locally stored schema entirely. After this operation the schema is equal to {@link DistributedSchema#EMPTY}.
      * The method does not execute any callback. It is indended to reinitialize the schema later using the method
      * {@link #reset(boolean)}.
+     *
+     * @return
```
Comment (adelapena) [inline src/java/org/apache/cassandra/schema/SchemaUpdateHandler.java]: I think we should either add some description to the `@return` tag or remove it.

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 28]
```diff
@@ -95,25 +107,54 @@ private void selectSilent(Cluster cluster, String name)
...
             cluster.get(2).shutdown().get();
 
-            // when schema is removed and there is no other node to fetch it from, node 1 should be left with clean schema
-            cluster.get(1).runOnInstance(() -> Schema.instance.resetLocalSchema());
+            Awaitility.await().atMost(Duration.ofSeconds(30)) // wait until node 1 notices that node 2 is dead
+                      .until(() -> cluster.get(1).callOnInstance(() -> Gossiper.instance.getLiveMembers().stream().allMatch(e -> e.equals(FBUtilities.getBroadcastAddressAndPort()))));
+
+            // when there is no node to fetch the schema from, reset local schema should immediately fail
+            Assertions.assertThatExceptionOfType(RuntimeException.class).isThrownBy(() -> {
+                cluster.get(1).runOnInstance(() -> Schema.instance.resetLocalSchema());
```
Comment (adelapena) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java:150]:
```suggestion
                cluster.get(1).runOnInstance(Schema.instance::resetLocalSchema);
```

Comment (adelapena) [reply test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java:150]: Please ignore this last suggestion, I forgot that method references don't work in the dtest framework.

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java (hunk 10 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Still trying to understand why it's needed, but for sure a quick attempt at removing it breaks the tests.

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: The same logic is in `pullComplete`.
Seems to me we `markReceived` before we deal with `requestQueue` in `maybePullSchema`, I am still not sure why though.... looking again into that

Comment (adelapena) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Actually, it hasn't necessarily marked are received, has it? It's only marked as received if it's equals to the local schema, I think.

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Yes, reading back that is right but then it is confusing why in those cases we add to the requestQueue which seems to be a queue of endpoints from which we are going to fetch.... do we need still to fetch? :-)

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Oh, that is covered in maybePullSchema, we check whether it was already marked received on the first line... It is just not immediately clear before you look into maybePullSchema... perhaps we can simplify or at least comment a bit...

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: At that point we  do not remove the endpoint from the queue, we seem to remove it next time we call maybePullSchema and we iterate the queue. Not sure why that approach was taken really and not just not adding it at all in case we markReceived... I will experiment

Comment (adelapena) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: So we add it to `requestQueue` because we might need to fetch it, unless it has been marked as received. I guess the simplified version would be adding the endpoint to the request queue only if hasn't been marked, something like:
```java
VersionInfo info = versionInfo.computeIfAbsent(version, VersionInfo::new);
info.endpoints.add(endpoint);
logger.trace("Added endpoint {} to schema {}: {}", endpoint, info.version, info);

if (Objects.equals(schemaVersion.get(), version))
{
    info.markReceived();
    logger.trace("Schema {} from {} has been marked as received because it is equal the local schema", version, endpoint);
}
else
{
    info.requestQueue.addFirst(endpoint);
}
```

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Yes, that was my point but I have to test if it can have some other side effects that we might miss...

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: At least the `MigrationCoordinatorTest` and `SchemaTest` (both in-jvm and unit tests) pass for me locally with this change

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java (hunk 28 above)
Comment (jacek-lewandowski) [reply test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java:150]: > Please ignore this last suggestion, I forgot that method references don't work in the dtest framework.

heh... yes, I struggled this problem some time ago. `runElsewhere(Schema.instance::resetLocalSchema)` means more or less:
```java
Schema schema = Schema.instance;
runElsewhere(() -> schema.resetLocalSchema());
```
so it is something completely different :-)

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java (hunk 4 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: I'll revert

Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: but actually I considered this message as temporary situation - when the messaging version is eventually known, the schema is either pulled or it is logged that the messaging version is incompatible (at debug level)

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java (hunk 5 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Ok, but where do you want it to be documented? I think we should rather fix it

File: src/java/org/apache/cassandra/schema/Schema.java (hunk 6 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/Schema.java:614]: Most general answer to that question is - because we are waiting in this method. The method simply mark a schema that is should be refreshed and wait until it happens. However, the schema would be unable to refresh if we hold a lock because we no changes could be applied then. Note that, the changes are applied from the MigrationCoordinator thread, usually a different to the called thread.

File: src/java/org/apache/cassandra/schema/SchemaUpdateHandler.java (hunk 27 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/SchemaUpdateHandler.java]: ok, and btw. I suppose the more adequate names for those methods now would be reset -> sync, clear -> refresh

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java (hunk 17 above)
Comment (jacek-lewandowski) [reply test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: I will, but with the current formatting rules I can see no way to format it nicely

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java (hunk 18 above)
Comment (jacek-lewandowski) [reply test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: again, the formatting rules does not allow for make it nicely to my knowledge

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java (hunk 12 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java:205]: we do not have to call `.toString()` explicitly in statements like `log.info("bla bla bla {}", info);` :)

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java (hunk 10 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: exactly, that was what I've commented - "...given we've just marked this schema version as received..." :-)

**Comment by jacek-lewandowski** [comment]
Thank you for thorough review, I've applied your suggestions

File: src/java/org/apache/cassandra/schema/MigrationCoordinator.java (hunk 5 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/MigrationCoordinator.java]: Then in a follow up ticket in order to be fixed and add the number in this comment? WDYT?

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 29]
```diff
@@ -93,70 +99,98 @@ private void selectSilent(Cluster cluster, String name)
...
     }
 
+    /**
+     * The purpose of this test is to verify manual schema reset functinality.
+     * <p>
+     * There is a 2-node cluster and a table one created. The schema version is agreed on both nodes. Then the 2nd node
+     * is shutdown, and we introduce a disagreement by dropping table one and creating table two. Therefore, 1st node
+     * has a newer schema version with table two, while the shutdown 2nd node has older schema version with table one.
+     * <p>
+     * At this point, if we just start 2nd node, it synces its schema by getting the fresh mutations from 1st node which
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: typo: `syncs`

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 30]
```diff
@@ -93,70 +99,98 @@ private void selectSilent(Cluster cluster, String name)
...
                 cause = cause.getCause();
             }
             assertTrue(causeIsUnknownColumn);
         }
     }
 
+    /**
+     * The purpose of this test is to verify manual schema reset functinality.
+     * <p>
+     * There is a 2-node cluster and a table one created. The schema version is agreed on both nodes. Then the 2nd node
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: I would use the names here probably `tbl_one` or so.

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 31]
```diff
@@ -93,70 +99,98 @@ private void selectSilent(Cluster cluster, String name)
...
+            assertTrue(checkTables(cluster.get(2), true, false));
         }
     }
 
+    private static ConditionFactory await(int seconds)
+    {
+        return Awaitility.await().atMost(ofSeconds(seconds)).pollDelay(ofSeconds(1));
+    }
+
+    private static boolean checkTables(IInvokableInstance instance, boolean one, boolean two)
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: I would probably change this name to `checkTablesPropagated` or something

File: test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java [hunk 32]
```diff
@@ -93,70 +99,98 @@ private void selectSilent(Cluster cluster, String name)
...
             }).withMessageContaining("Cannot reset local schema when there are no other live nodes");
 
-            // now, let's make a disagreement, the shutdown node 2 has a definition of tbl, while the running node 1 does not
-            cluster.get(1).runOnInstance(() -> {
-                Schema.instance.transform(current -> Schema.instance.distributedKeyspaces().without(KEYSPACE), false);
-             });
-
-            assertTrue(cluster.get(1).callOnInstance(() -> Schema.instance.getTableMetadata(KEYSPACE, "tbl") == null));
+            // now, let's make a disagreement, the shutdown node 2 has a definition of table one, while
+            // the running node 2 will have definitions of table two
```
Comment (ekaterinadimitrova2) [inline test/distributed/org/apache/cassandra/distributed/test/SchemaTest.java]: `a definition`

**Comment by jacek-lewandowski** [comment]
Merged to cassandra-4.1 in d8bbeb9e398723e5efe13166e16372d104fb93ce
